From b6fca0f36a60902d562e0725d11d16483838962f Mon Sep 17 00:00:00 2001
Message-Id: <b6fca0f36a60902d562e0725d11d16483838962f.1764249235.git.gerbach@fs-net.de>
In-Reply-To: <58ad0b7562e0114082e19f98da1053f1896ae596.1764249234.git.gerbach@fs-net.de>
References: <58ad0b7562e0114082e19f98da1053f1896ae596.1764249234.git.gerbach@fs-net.de>
From: Philipp Gerbach <gerbach@fs-net.de>
Date: Tue, 13 May 2025 16:49:07 +0200
Subject: [PATCH 05/11] Improve rtc-pcf85363 driver

This patch adds additional device-tree configuration options, like
output clock speed, quarz drive strength and quartz low jitter.

We use them on most newer boards.

We should try to get this mainline.
---
 drivers/rtc/rtc-pcf85363.c | 90 ++++++++++++++++++++++++++++++++++++--
 1 file changed, 86 insertions(+), 4 deletions(-)

diff --git a/drivers/rtc/rtc-pcf85363.c b/drivers/rtc/rtc-pcf85363.c
index 540042b9e..4bb4d2200 100644
--- a/drivers/rtc/rtc-pcf85363.c
+++ b/drivers/rtc/rtc-pcf85363.c
@@ -103,6 +103,26 @@
 #define OSC_CAP_SEL	GENMASK(1, 0)
 #define OSC_CAP_6000	0x01
 #define OSC_CAP_12500	0x02
+#define OSC_CAP_SHIFT	0
+
+#define OSC_DRIVE_SEL	GENMASK(3, 2)
+#define OSC_DRIVE_60	0x01
+#define OSC_DRIVE_500	0x02
+#define OSC_DRIVE_SHIFT	2
+
+#define OSC_LOWJ		BIT(4)
+#define OSC_12H			BIT(5)
+
+#define REG_FUNCTION	0x28
+#define COF_SEL		GENMASK(2, 0)
+#define COF_32kHz	0x00	/* clock output 32,768 kHz */
+#define COF_16kHz	0x01	/* clock output 16,384 kHz */
+#define COF_8kHz	0x02	/* clock output 8,192 kHz */
+#define COF_4kHz	0x03	/* clock output 4,096 kHz */
+#define COF_2kHz	0x04	/* clock output 2,048 kHz */
+#define COF_1kHz	0x05	/* clock output 1,024 kHz */
+#define COF_1Hz	0x06	/* clock output 1 Hz */
+#define COF_OFF	0x07	/* No clock output */
 
 #define STOP_EN_STOP	BIT(0)
 
@@ -122,7 +142,10 @@ struct pcf85x63_config {
 
 static int pcf85363_load_capacitance(struct pcf85363 *pcf85363, struct device_node *node)
 {
+	int ret;
 	u32 load = 7000;
+	u32 drive = 100;
+	u32 clk_out = 32768;
 	u8 value = 0;
 
 	of_property_read_u32(node, "quartz-load-femtofarads", &load);
@@ -135,15 +158,74 @@ static int pcf85363_load_capacitance(struct pcf85363 *pcf85363, struct device_no
 	case 7000:
 		break;
 	case 6000:
-		value = OSC_CAP_6000;
+		value = (OSC_CAP_6000 & OSC_CAP_SEL);
 		break;
 	case 12500:
-		value = OSC_CAP_12500;
+		value = (OSC_CAP_12500 & OSC_CAP_SEL);
+		break;
+	}
+
+	of_property_read_u32(node, "quartz-drive-kiloohm", &drive);
+	switch (drive) {
+	default:
+		dev_warn(&pcf85363->rtc->dev, "Unknown quartz-drive-kiloohm value: %d. Assuming 100",
+			 drive);
+		fallthrough;
+	case 100:
+		break;
+	case 60:
+		value |= ((OSC_DRIVE_60 << OSC_DRIVE_SHIFT)
+		    & OSC_DRIVE_SEL);
+		break;
+	case 500:
+		value |= ((OSC_DRIVE_500 << OSC_DRIVE_SHIFT)
+		    & OSC_DRIVE_SEL);
+		break;
+	}
+
+	if (of_property_read_bool(node, "quartz-low-jitter"))
+		value |= OSC_LOWJ;
+
+	return regmap_write(pcf85363->regmap, CTRL_OSCILLATOR, value);
+	if (ret)
+		return ret;
+
+	of_property_read_u32(node, "colck-out-Hz", &clk_out);
+	/* Set clock output frequency if available */
+	switch (clk_out) {
+	default:
+		dev_warn(&pcf85363->rtc->dev, "Unknown clk-out-Hz value: %d. Assuming 32768",
+			 drive);
+		fallthrough;
+	case 32768:
+		value = COF_32kHz;
+		break;
+	case 6384:
+		value = COF_16kHz;
+		break;
+	case 8192:
+		value = COF_8kHz;
+		break;
+	case 4096:
+		value = COF_4kHz;
 		break;
+	case 2048:
+		value = COF_2kHz;
+		break;
+	case 1024:
+		value = COF_1kHz;
+		break;
+	case 1:
+		value = COF_1Hz;
+		break;
+	case 0:
+		value = COF_OFF;
+
 	}
+	ret = regmap_update_bits(pcf85363->regmap, REG_FUNCTION, COF_SEL, value);
+
 
-	return regmap_update_bits(pcf85363->regmap, CTRL_OSCILLATOR,
-				  OSC_CAP_SEL, value);
+	return ret;
 }
 
 static int pcf85363_rtc_read_time(struct device *dev, struct rtc_time *tm)
-- 
2.40.1

